prow/cmd/hook/main.go

```diff
@@ -220,5 +220,20 @@ func main() {
 	// Serve plugin help information from /plugin-help.
 	http.Handle("/plugin-help", pluginhelp.NewHelpAgent(pluginAgent, githubClient))
 
-	logrus.Fatal(http.ListenAndServe(":"+strconv.Itoa(o.port), nil))
+	httpServer := &http.Server{Addr: ":" + strconv.Itoa(o.port), Handler: nil}
+
+	// Shutdown gracefully on SIGTERM or SIGINT
+	go func() {
+		sig := make(chan os.Signal, 1)
+		signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
+		select {
+		case <-sig:
+			logrus.Info("Hook is shutting down...")
+			ctx, cancel := context.WithTimeout(context.Background(), o.gracePeriod)
+			defer cancel()
+			httpServer.Shutdown(ctx)
+		}
+	}()
+
+	logrus.Fatal(httpServer.ListenAndServe())
```

Comment:
I don't think it is safe to wrap `ListenAndServe` in `Fatal` now. From the docs for `Server.Shutdown`:
>When Shutdown is called, Serve, ListenAndServe, and ListenAndServeTLS immediately return ErrServerClosed. Make sure the program doesn't exit and waits instead for Shutdown to return.

Also keep in mind that `Fatal` calls `os.Exit(1)` which will bypass the defered `server.GracefulShutdown()`.

Reply:
this should probably be something more like `logrus.WithError(httpServer.ListenAndServe()).Warn("Server exited")`

Reply:
Should be fixed in 72cddfc